lib/spack/spack/build_systems/go.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import inspect
+
+import llnl.util.filesystem as fs
+
+import spack.builder
+import spack.package_base
+from spack.directives import build_system, extends
+from spack.multimethod import when
+
+from ._checks import BaseBuilder, execute_install_time_tests
+
+
+class GoPackage(spack.package_base.PackageBase):
+    """Specialized class for packages built using the Go toolchain."""
+
+    #: This attribute is used in UI queries that need to know the build
+    #: system base class
+    build_system_class = "GoPackage"
+
+    #: Legacy buildsystem attribute used to deserialize and install old specs
+    legacy_buildsystem = "go"
+
+    build_system("go")
+
+    with when("build_system=go"):
+        # TODO: this seems like it should be depends_on, see
+        # setup_dependent_build_environment in go for why I kept it like this
+        extends("go@1.14:", type="build")
+
+
+@spack.builder.builder("go")
+class GoBuilder(BaseBuilder):
+    """The Go builder encodes the most common way of building software with
+    a golang go.mod file. It has two phases that can be overridden, if need be:
+
+            1. :py:meth:`~.GoBuilder.build`
+            2. :py:meth:`~.GoBuilder.install`
+
+    For a finer tuning you may override:
+
+        +-----------------------------------------------+--------------------+
+        | **Method**                                    | **Purpose**        |
+        +===============================================+====================+
+        | :py:meth:`~.GoBuilder.build_args`             | Specify arguments  |
+        |                                               | to ``go build``    |
+        +-----------------------------------------------+--------------------+
+        | :py:meth:`~.GoBuilder.test_args`              | Specify arguments  |
+        |                                               | to ``go test``     |
+        +-----------------------------------------------+--------------------+
+    """
+
+    phases = ("build", "install")
+
+    #: Callback names for install-time test
+    install_time_test_callbacks = ["check"]
+
+    def setup_build_environment(self, env):
+        env.set("GO111MODULE", "on")
+        env.set("GOTOOLCHAIN", "local")
+
+    @property
+    def build_directory(self):
+        """Return the directory containing the main go.mod."""
+        return self.pkg.stage.source_path
+
+    @property
+    def build_args(self):
+        """Arguments for ``go build``."""
+        return ["-ldflags", "-s -w", "-o", f"{self.pkg.name}"]
```

Comment:
Do we need a doc line here? Or use full options?
```suggestion
        # Always pass ldflags -s = --strip-all and -w = --no-warnings
        return ["-ldflags", "-s -w", "-o", f"{self.pkg.name}"]
```

Reply:
Actually, that makes me wonder.  The no-warnings one makes perfect sense, why `--strip-all`?

Reply:
I guess it's intended to mimic the CMake Release build type, as opposed to RelWithDebInfo without `-s`. Since Release is the default for CMake, it makes sense to make `-s` the default here. But should we then allow `ldflags`, `gcflags`, and `gccgoflags` much like `cflags`?

Reply:
Yep, @wdconinc got it. I opted to use `-s` here to mimic CMake or Rust release builds and strip out symbols and debugging info. I think adding a comment explaining the flags is a great idea. If folks want to change the defaults it'll be pretty easy to define their own args in a package so I'm not too worried about changing that here.

Reply:
Makes sense.  The flags I wouldn't worry about _yet_ because the generic mechanism for handling flags for non-c/c++/fortran compilers wont quite be feasible until we have the other bits of compilers as deps, so we can make the go compiler a "real compiler" in terms of spack's internals.  If we find we really need it for some reason we could fake it with a variant but it wouldn't behave the same.